// @sparted/eslint-plugin/lib/configs/plugins/jest/common.js
'use strict';


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


module.exports = {
  env: {
    '@sparted/jest/globals': true,
  },
  rules: {
    '@sparted/jest/consistent-test-it': [ERROR, {
      fn: 'test', // Default
      withinDescribe: 'it', // Default
    }],
    '@sparted/jest/expect-expect': [ERROR, {
      assertFunctionNames: ['expect'], // Default
    }],
    '@sparted/jest/lowercase-name': [ERROR, {
      ignore: [], // Default
      allowedPrefixes: [], // Default
      ignoreTopLevelDescribe: true,
    }],
    '@sparted/jest/no-alias-methods': [ERROR],
    '@sparted/jest/no-commented-out-tests': [ERROR],
    '@sparted/jest/no-conditional-expect': [ERROR],
    '@sparted/jest/no-deprecated-functions': [ERROR],
    // Only WARN because it is better to have a skipped test than no test at all
    '@sparted/jest/no-disabled-tests': [WARN],
    '@sparted/jest/no-done-callback': [ERROR],
    '@sparted/jest/no-duplicate-hooks': [ERROR],
    '@sparted/jest/no-export': [ERROR],
    '@sparted/jest/no-focused-tests': [ERROR],
    // Could be debatable
    '@sparted/jest/no-hooks': [ERROR, {
      // Default
      // allow: [],
    }],
    '@sparted/jest/no-identical-title': [ERROR],
    '@sparted/jest/no-if': [ERROR],
    '@sparted/jest/no-interpolation-in-snapshots': [ERROR],
    '@sparted/jest/no-jasmine-globals': [ERROR],
    '@sparted/jest/no-jest-import': [ERROR],
    '@sparted/jest/no-large-snapshots': [ERROR, {
      maxSize: 50, // Default
      inlineMaxSize: 50, // Default
      allowedSnapshots: {}, // Default
    }],
    '@sparted/jest/no-mocks-import': [ERROR],
    // Should be populated with unwanted matchers
    '@sparted/jest/no-restricted-matchers': [ERROR],
    '@sparted/jest/no-standalone-expect': [ERROR, {
      additionalTestBlockFunctions: [], // Default
    }],
    '@sparted/jest/no-test-prefixes': [ERROR],
    '@sparted/jest/no-test-return-statement': [ERROR],
    '@sparted/jest/prefer-called-with': [ERROR],
    '@sparted/jest/prefer-expect-assertions': [ERROR, {
      onlyFunctionsWithAsyncKeyword: true,
    }],
    '@sparted/jest/prefer-hooks-on-top': [ERROR],
    '@sparted/jest/prefer-spy-on': [ERROR],
    '@sparted/jest/prefer-strict-equal': [ERROR],
    '@sparted/jest/prefer-to-be-null': [ERROR],
    '@sparted/jest/prefer-to-be-undefined': [ERROR],
    '@sparted/jest/prefer-to-contain': [ERROR],
    '@sparted/jest/prefer-to-have-length': [ERROR],
    '@sparted/jest/prefer-todo': [ERROR],
    '@sparted/jest/require-to-throw-message': [ERROR],
    '@sparted/jest/require-top-level-describe': [ERROR],
    '@sparted/jest/valid-describe': [ERROR],
    '@sparted/jest/valid-expect': [ERROR, {
      alwaysAwait: true,
      minArgs: 1, // Default
      maxArgs: 1, // Default
    }],
    '@sparted/jest/valid-expect-in-promise': [ERROR],
    '@sparted/jest/valid-title': [ERROR, {
      ignoreTypeOfDescribeName: false, // Default
      disallowedWords: [], // Default
      mustNotMatch: {}, // Default
      mustMatch: {
        test: '^that',
        it: '^should',
      },
    }],
  },
};
