// eslint-plugin-jsonc/dist/rules/valid-json-number.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsonc_eslint_parser_1 = require("jsonc-eslint-parser");
const utils_1 = require("../utils");
function isValidNumber(text) {
    if (text.startsWith(".") || text.endsWith(".")) {
        return false;
    }
    try {
        JSON.parse(text);
    }
    catch (_a) {
        return false;
    }
    return true;
}
exports.default = utils_1.createRule("valid-json-number", {
    meta: {
        docs: {
            description: "disallow invalid number for JSON",
            recommended: ["json", "jsonc"],
            extensionRule: false,
        },
        fixable: "code",
        schema: [],
        messages: {
            invalid: "Invalid number for JSON.",
            invalidSpace: "Spaces after minus sign are not allowed in JSON.",
            invalidPlus: "Plus signs are not allowed in JSON.",
            invalidIdentifier: "`{{name}}` are not allowed in JSON.",
        },
        type: "problem",
    },
    create(context) {
        if (!context.parserServices.isJSON) {
            return {};
        }
        const sourceCode = context.getSourceCode();
        return {
            JSONUnaryExpression(node) {
                if (node.argument.type === "JSONIdentifier") {
                    return;
                }
                const operator = sourceCode.getFirstToken(node, (token) => token.type === "Punctuator" &&
                    token.value === node.operator);
                if (node.operator === "+") {
                    context.report({
                        loc: (operator === null || operator === void 0 ? void 0 : operator.loc) || node.loc,
                        messageId: "invalidPlus",
                        fix(fixer) {
                            return operator ? fixer.remove(operator) : null;
                        },
                    });
                }
                else if (operator &&
                    operator.range[1] < node.argument.range[0]) {
                    context.report({
                        loc: {
                            start: operator.loc.end,
                            end: node.argument.loc.start,
                        },
                        messageId: "invalidSpace",
                        fix(fixer) {
                            return fixer.removeRange([
                                operator.range[1],
                                node.argument.range[0],
                            ]);
                        },
                    });
                }
            },
            JSONLiteral(node) {
                if (typeof node.value !== "number") {
                    return;
                }
                const text = sourceCode.text.slice(...node.range);
                if (!isValidNumber(text)) {
                    context.report({
                        loc: node.loc,
                        messageId: "invalid",
                        fix(fixer) {
                            return fixer.replaceTextRange(node.range, `${node.value}`);
                        },
                    });
                }
            },
            JSONIdentifier(node) {
                if (!jsonc_eslint_parser_1.isNumberIdentifier(node)) {
                    return;
                }
                context.report({
                    loc: node.loc,
                    messageId: "invalidIdentifier",
                    data: {
                        name: node.name,
                    },
                });
            },
        };
    },
});
