// @sparted/eslint-plugin/lib/configs/plugins/jsx-a11y/common.js
'use strict';

const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


const labelAttributes = ['label']; // Default
const controlComponents = []; // Default
const imageComponents = ['Image', 'Picture'];
// Default
const interactiveHandlers = [
  'onClick',
  'onMouseDown',
  'onMouseUp',
  'onKeyPress',
  'onKeyDown',
  'onKeyUp',
];


module.exports = {
  rules: {
    '@sparted/jsx-a11y/accessible-emoji': [ERROR],
    '@sparted/jsx-a11y/alt-text': [ERROR, {
      elements: ['img', 'object', 'area', 'input[type="image"]'], // Default
      img: imageComponents,
      object: ['Object'],
      area: ['Area'],
      'input[type="image"]': ['InputImage'],
    }],
    '@sparted/jsx-a11y/anchor-has-content': [ERROR, {
      components: ['Link'],
    }],
    '@sparted/jsx-a11y/anchor-is-valid': [ERROR, {
      components: ['Link'],
      specialLink: ['href', 'url'],
      aspects: ['noHref', 'invalidHref', 'preferButton'], // Default
    }],
    '@sparted/jsx-a11y/aria-activedescendant-has-tabindex': [ERROR],
    '@sparted/jsx-a11y/aria-props': [ERROR],
    '@sparted/jsx-a11y/aria-proptypes': [ERROR],
    '@sparted/jsx-a11y/aria-role': [ERROR],
    '@sparted/jsx-a11y/aria-unsupported-elements': [ERROR],
    '@sparted/jsx-a11y/autocomplete-valid': [ERROR, {
      inputComponents: ['input', 'Input', 'Field'],
    }],
    '@sparted/jsx-a11y/click-events-have-key-events': [ERROR],
    '@sparted/jsx-a11y/control-has-associated-label': [ERROR, {
      labelAttributes,
      controlComponents,
      ignoreElements: [], // Default
      ignoreRoles: [], // Default
      depth: 2, // Default
    }],
    '@sparted/jsx-a11y/heading-has-content': [ERROR, {
      components: ['Heading'],
    }],
    '@sparted/jsx-a11y/html-has-lang': [ERROR],
    '@sparted/jsx-a11y/iframe-has-title': [ERROR],
    '@sparted/jsx-a11y/img-redundant-alt': [ERROR, {
      components: imageComponents,
      words: ['image', 'photo', 'picture'],
    }],
    '@sparted/jsx-a11y/interactive-supports-focus': [ERROR, {
      // Default
      tabbable: [
        'button',
        'checkbox',
        'link',
        'searchbox',
        'spinbutton',
        'switch',
        'textbox',
      ],
    }],
    '@sparted/jsx-a11y/label-has-associated-control': [ERROR, {
      labelComponents: ['Label'],
      labelAttributes,
      controlComponents,
      assert: 'both',
      depth: 2, // Default
    }],
    '@sparted/jsx-a11y/lang': [ERROR],
    '@sparted/jsx-a11y/media-has-caption': [ERROR, {
      audio: ['Audio'],
      video: ['Video'],
      track: ['Track'],
    }],
    '@sparted/jsx-a11y/mouse-events-have-key-events': [ERROR],
    '@sparted/jsx-a11y/no-access-key': [ERROR],
    // OFF as autofocus could be a major UX improvements
    '@sparted/jsx-a11y/no-autofocus': [OFF, {
      ignoreNonDOM: false, // Default
    }],
    '@sparted/jsx-a11y/no-distracting-elements': [ERROR, {
      elements: ['marquee', 'blink'],
    }],
    '@sparted/jsx-a11y/no-interactive-element-to-noninteractive-role': [ERROR, {
      tr: ['none', 'presentation'],
    }],
    '@sparted/jsx-a11y/no-noninteractive-element-interactions': [ERROR, {
      // Default
      handlers: interactiveHandlers,
    }],
    '@sparted/jsx-a11y/no-noninteractive-element-to-interactive-role': [ERROR, {
      ul: ['listbox', 'menu', 'menubar', 'radiogroup', 'tablist', 'tree', 'treegrid'],
      ol: ['listbox', 'menu', 'menubar', 'radiogroup', 'tablist', 'tree', 'treegrid'],
      li: ['menuitem', 'option', 'row', 'tab', 'treeitem'],
      table: ['grid'],
      td: ['gridcell'],
    }],
    '@sparted/jsx-a11y/no-noninteractive-tabindex': [ERROR, {
      tags: [],
      roles: ['tabpanel'],
    }],
    '@sparted/jsx-a11y/no-onchange': [ERROR],
    '@sparted/jsx-a11y/no-redundant-roles': [ERROR, {
      nav: ['navigation'],
    }],
    '@sparted/jsx-a11y/no-static-element-interactions': [ERROR, {
      // Default
      handlers: interactiveHandlers,
    }],
    '@sparted/jsx-a11y/role-has-required-aria-props': [ERROR],
    '@sparted/jsx-a11y/role-supports-aria-props': [ERROR],
    '@sparted/jsx-a11y/scope': [ERROR],
    '@sparted/jsx-a11y/tabindex-no-positive': [ERROR],
  },
};
