// @sparted/eslint-plugin/lib/configs/plugins/unicorn/common.js
'use strict';


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


module.exports = {
  rules: {
    '@sparted/unicorn/better-regex': [ERROR, {
      sortCharacterClasses: true, // Default
    }],
    '@sparted/unicorn/catch-error-name': [ERROR, {
      name: 'error', // Default
      ignore: [], // Default
    }],
    '@sparted/unicorn/consistent-function-scoping': [ERROR],
    '@sparted/unicorn/custom-error-definition': [ERROR],
    '@sparted/unicorn/error-message': [ERROR],
    '@sparted/unicorn/escape-case': [ERROR],
    // Related to the core rule no-warning-comments
    '@sparted/unicorn/expiring-todo-comments': [ERROR, {
      ignoreDatesOnPullRequests: true, // Default
      terms: ['todo', 'fixme'],
      allowWarningComments: true, // Default
      ignore: [], // Default
    }],
    // Debatable, the usage might prevail
    '@sparted/unicorn/explicit-length-check': [ERROR, {
      'non-zero': 'greater-than', // Default
    }],
    '@sparted/unicorn/filename-case': [ERROR, {
      case: 'kebabCase',
      ignore: [
        /\.md$/ui,
      ],
    }],
    // Might cause issue in node context
    '@sparted/unicorn/import-index': [ERROR, {
      ignoreImports: false, // Default
    }],
    '@sparted/unicorn/import-style': [ERROR, {
      // Default
      styles: {
        util: { named: true },
        path: { default: true },
        chalk: { default: true },
      },
      extendDefaultStyles: true, // Default
      checkImport: true, // Default
      checkDynamicImport: true, // Default
      checkExportFrom: false, // Default
      checkRequire: true, // Default
    }],
    // Related to the core rule no-new-wrappers
    '@sparted/unicorn/new-for-builtins': [ERROR],
    // TODO: Might be duplicate of the rule eslint-comments/no-unlimited-disable
    '@sparted/unicorn/no-abusive-eslint-disable': [ERROR],
    '@sparted/unicorn/no-array-instanceof': [ERROR],
    '@sparted/unicorn/no-console-spaces': [ERROR],
    '@sparted/unicorn/no-fn-reference-in-iterator': [ERROR],
    '@sparted/unicorn/no-for-loop': [ERROR],
    '@sparted/unicorn/no-hex-escape': [ERROR],
    '@sparted/unicorn/no-keyword-prefix': [ERROR, {
      blacklist: ['new', 'class'], // Default
      checkProperties: false,
      onlyCamelCase: true, // Default
    }],
    // Supersed the core rule no-nested-ternary
    '@sparted/unicorn/no-nested-ternary': [ERROR],
    // OFF as it is already handle by the rule node/no-deprecated-api
    '@sparted/unicorn/no-new-buffer': [OFF],
    // OFF as it is too restrictive
    '@sparted/unicorn/no-null': [OFF, {
      checkStrictEquality: true,
    }],
    '@sparted/unicorn/no-object-as-default-parameter': [ERROR],
    // Supersed the core rule no-process-exit
    '@sparted/unicorn/no-process-exit': [ERROR],
    // OFF as we are not ready for that
    '@sparted/unicorn/no-reduce': [OFF],
    '@sparted/unicorn/no-unreadable-array-destructuring': [ERROR],
    '@sparted/unicorn/no-unsafe-regex': [ERROR],
    '@sparted/unicorn/no-unused-properties': [ERROR],
    // TODO: Might conflict with the core rule array-callback-return
    '@sparted/unicorn/no-useless-undefined': [ERROR, {
      checkArguments: true, // Default
    }],
    '@sparted/unicorn/no-zero-fractions': [ERROR],
    '@sparted/unicorn/number-literal-case': [ERROR],
    '@sparted/unicorn/numeric-separators-style': [ERROR, {
      hexadecimal: {
        minimumDigits: 0,
        groupLength: 2,
      },
      binary: {
        minimumDigits: 0,
        groupLength: 4,
      },
      octal: {
        minimumDigits: 0,
        groupLength: 4,
      },
      number: {
        minimumDigits: 5,
        groupLength: 3,
      },
    }],
    '@sparted/unicorn/prefer-array-find': [ERROR],
    '@sparted/unicorn/prefer-flat-map': [ERROR],
    '@sparted/unicorn/prefer-includes': [ERROR],
    '@sparted/unicorn/prefer-math-trunc': [ERROR],
    '@sparted/unicorn/prefer-negative-index': [ERROR],
    '@sparted/unicorn/prefer-number-properties': [ERROR, {
      checkInfinity: false,
    }],
    '@sparted/unicorn/prefer-optional-catch-binding': [ERROR],
    '@sparted/unicorn/prefer-reflect-apply': [ERROR],
    '@sparted/unicorn/prefer-replace-all': [ERROR],
    '@sparted/unicorn/prefer-set-has': [ERROR],
    // Related to the core rule prefer-spread
    '@sparted/unicorn/prefer-spread': [ERROR],
    '@sparted/unicorn/prefer-starts-ends-with': [ERROR],
    '@sparted/unicorn/prefer-string-slice': [ERROR],
    '@sparted/unicorn/prefer-ternary': [ERROR],
    '@sparted/unicorn/prefer-text-content': [ERROR],
    '@sparted/unicorn/prefer-trim-start-end': [ERROR],
    '@sparted/unicorn/prefer-type-error': [ERROR],
    // Should be configured according to specific project standard
    '@sparted/unicorn/prevent-abbreviations': [ERROR, {
      replacements: {}, // Default here: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/prevent-abbreviations.js#L18
      extendDefaultReplacements: true, // Default
      // Default here: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/prevent-abbreviations.js#L222
      whitelist: {
        i: true,
        str: true,
        src: true,
        dest: true,
      },
      extendDefaultWhitelist: true, // Default
      checkDefaultAndNamespaceImports: 'internal', // Default
      checkShorthandImports: 'internal', // Default
      checkShorthandProperties: false, // Default
      checkProperties: false, // Default
      checkVariables: true, // Default
      checkFilenames: true, // Default
    }],
    '@sparted/unicorn/string-content': [ERROR, {
      patterns: {
        '^http:\\/\\/': {
          suggest: '^https:\\/\\/',
          message: 'Please secure your links or disable the rule on this line',
        },
      },
    }],
    '@sparted/unicorn/throw-new-error': [ERROR],

    /* ----- Reset base config ----- */
    // Superseded by unicorn/no-nested-ternary
    'no-nested-ternary': [OFF],
    // Superseded by unicorn/no-process-exit
    'no-process-exit': [OFF],
  },
};
