// @sparted/eslint-plugin/lib/configs/plugins/shopify/+jsx.js
'use strict';


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';

const ARIA_LABEL = 'aria-label';


module.exports = {
  rules: {
    '@sparted/shopify/jsx-no-complex-expressions': [ERROR],
    '@sparted/shopify/jsx-no-hardcoded-content': [ERROR, {
      allowStrings: false, // Default
      allowNumbers: false,
      checkProps: [], // Default
      dom: {
        '*': {
          allowStrings: false, // Default
          allowNumbers: false,
          checkProps: ['title', ARIA_LABEL],
        },
        area: {
          allowStrings: false, // Default
          allowNumbers: false,
          checkProps: ['title', 'alt', ARIA_LABEL],
        },
        img: {
          allowStrings: false, // Default
          allowNumbers: false,
          checkProps: ['title', 'alt', ARIA_LABEL],
        },
        input: {
          allowStrings: false, // Default
          allowNumbers: false,
          checkProps: ['title', 'alt', 'placeholder', ARIA_LABEL],
        },
      },
      modules: {}, // Default
    }],
    '@sparted/shopify/jsx-prefer-fragment-wrappers': [ERROR],
  },
};
