// @sparted/eslint-plugin/lib/configs/plugins/shopify/common.js
'use strict';


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


module.exports = {
  rules: {
    '@sparted/shopify/binary-assignment-parens': [ERROR, 'always'],
    '@sparted/shopify/class-property-semi': [ERROR, 'always'],
    '@sparted/shopify/images-no-direct-imports': [ERROR],
    '@sparted/shopify/no-ancestor-directory-import': [ERROR],
    // OFF as it is a duplicate of the core rule no-debugger
    '@sparted/shopify/no-debugger': [OFF],
    '@sparted/shopify/no-fully-static-classes': [ERROR],
    // OFF as it conlict with the rule import/no-namespace
    // As this rule have an option to allow specific modules
    // unlike import/no-namespace, this two rules might be inverted
    '@sparted/shopify/no-namespace-imports': [OFF, {
      allow: [], // Default
    }],
    '@sparted/shopify/no-useless-computed-properties': [ERROR],
    // OFF as we don't use Polaris: https://github.com/Shopify/polaris-react
    '@sparted/shopify/polaris-no-bare-stack-item': [OFF],
    // OFF as we don't use Polaris: https://github.com/Shopify/polaris-react
    '@sparted/shopify/polaris-prefer-sectioned-prop': [OFF],
    // OFF as it is buggy, doesn't seems to support spread operator in this.state
    // TODO: fix it
    '@sparted/shopify/prefer-class-properties': [OFF, 'always'],
    '@sparted/shopify/prefer-early-return': [ERROR, {
      maximumStatements: 1, // Default
    }],
    '@sparted/shopify/prefer-module-scope-constants': [ERROR],
    // OFF as we don't use Twine: https://github.com/Shopify/twine
    '@sparted/shopify/prefer-twine': [OFF],
    '@sparted/shopify/restrict-full-import': [ERROR, []], // Default
    // OFF as we don't use sinon: https://github.com/sinonjs/sinon
    '@sparted/shopify/sinon-no-restricted-features': [OFF, {
      restricted: [], // Default
      aliases: ['sinon'], // Default
      injected: true,
    }],
    // OFF as we don't use sinon: https://github.com/sinonjs/sinon
    '@sparted/shopify/sinon-prefer-meaningful-assertions': [OFF],
    // TODO: Might conflict with the plugin eslint-plugin-boundaries
    '@sparted/shopify/strict-component-boundaries': [ERROR],
  },
};
