// @sparted/eslint-plugin/lib/configs/plugins/flowtype/common.js
'use strict';

const { getRuleConfig } = require('../../../utils.js');
const baseConfig = require('../core/common.js');


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';

const ALWAYS_MULTILINE = 'always-multiline';


module.exports = {
  parser: '@babel/eslint-parser',
  settings: {
    flowtype: {
      onlyFilesWithFlowAnnotation: true,
    },
  },
  rules: {
    '@sparted/flowtype/array-style-complex-type': [ERROR, 'verbose'], // Default
    '@sparted/flowtype/array-style-simple-type': [ERROR, 'verbose'], // Default
    '@sparted/flowtype/arrow-parens': [ERROR, 'always', {
      requireForBlockBody: false, // Default
    }],
    '@sparted/flowtype/boolean-style': [ERROR, 'boolean'], // Default
    '@sparted/flowtype/define-flow-type': [ERROR],
    // Related to comma-dangle
    '@sparted/flowtype/delimiter-dangle': [ERROR, ALWAYS_MULTILINE, ALWAYS_MULTILINE, ALWAYS_MULTILINE],
    '@sparted/flowtype/generic-spacing': [ERROR, 'never'],
    '@sparted/flowtype/newline-after-flow-annotation': [ERROR, 'always'],
    '@sparted/flowtype/no-dupe-keys': [ERROR],
    '@sparted/flowtype/no-existential-type': [ERROR],
    // WARN only to fix it over time
    '@sparted/flowtype/no-flow-fix-me-comments': [WARN],
    '@sparted/flowtype/no-mixed': [ERROR],
    // Might be an issue to store an array of refs
    '@sparted/flowtype/no-mutable-array': [ERROR],
    '@sparted/flowtype/no-primitive-constructor-types': [ERROR],
    '@sparted/flowtype/no-types-missing-file-annotation': [ERROR],
    // Supersed no-unused-expressions
    '@sparted/flowtype/no-unused-expressions': getRuleConfig('no-unused-expressions', baseConfig),
    '@sparted/flowtype/no-weak-types': [ERROR, {
      any: true, // Default
      Object: true, // Default
      Function: true, // Default
    }],
    '@sparted/flowtype/object-type-delimiter': [ERROR, 'comma'], // Default
    '@sparted/flowtype/require-compound-type-alias': [ERROR, 'always'], // Default
    '@sparted/flowtype/require-exact-type': [ERROR, 'always'], // Default
    // Might be problematic to declare generic object type
    '@sparted/flowtype/require-indexer-name': [ERROR, 'always'],
    '@sparted/flowtype/require-inexact-type': [ERROR, 'always'],
    '@sparted/flowtype/require-parameter-type': [ERROR, {
      excludeArrowFunctions: false, // Default
      excludeParameterMatch: '^_$',
    }],
    '@sparted/flowtype/require-readonly-react-props': [ERROR],
    '@sparted/flowtype/require-return-type': [ERROR, 'always', {
      excludeArrowFunctions: 'expressionsOnly',
      annotateUndefined: 'always',
      excludeMatching: ['constructor'],
    }],
    '@sparted/flowtype/require-types-at-top': [ERROR, 'always'],
    '@sparted/flowtype/require-valid-file-annotation': [ERROR, 'always', {
      annotationStyle: 'line',
      strict: false, // Default
    }],
    '@sparted/flowtype/require-variable-type': [ERROR, {
      excludeVariableMatch: 'a^', // Default
      excludeVariableTypes: {
        var: false, // Default
        let: false, // Default
        const: false, // Default
      },
    }],
    '@sparted/flowtype/semi': [ERROR, 'always'],
    // OFF as it is a nightmare to maintain for little benefit
    '@sparted/flowtype/sort-keys': [OFF, 'asc', {
      caseSensitive: false,
      natural: true,
    }],
    '@sparted/flowtype/space-after-type-colon': [ERROR, 'always', {
      allowLineBreak: false,
    }],
    '@sparted/flowtype/space-before-generic-bracket': [ERROR, 'never'],
    '@sparted/flowtype/space-before-type-colon': [ERROR, 'never'],
    '@sparted/flowtype/spread-exact-type': [ERROR],
    // Start with an uppercase, in pascal case, without ending by `Type`
    '@sparted/flowtype/type-id-match': [ERROR, '/^(?:[A-Z][a-z0-9]*)+(?<!Type)$/'],
    '@sparted/flowtype/type-import-style': [ERROR, 'declaration', {
      ignoreTypeDefault: false, // Default
    }],
    '@sparted/flowtype/union-intersection-spacing': [ERROR, 'always'],
    // Prevent conflict with the core rule no-unused-vars
    '@sparted/flowtype/use-flow-type': [ERROR],
  },
};
