// @sparted/eslint-plugin/lib/configs/plugins/node/common.js
'use strict';

const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


module.exports = {
  settings: {
    node: {
      // Used by the rules:
      // - no-extraneous-import
      // - no-extraneous-require
      // - no-missing-import
      // - no-missing-require
      allowModules: [], // Default
      resolvePaths: [__dirname],
      tryExtensions: ['.js', '.json'],
    },
  },
  rules: {
    /* ----- Possible errors ----- */
    '@sparted/node/handle-callback-err': [ERROR, 'error'],
    '@sparted/node/no-callback-literal': [ERROR],
    '@sparted/node/no-exports-assign': [ERROR],
    '@sparted/node/no-extraneous-import': [ERROR],
    '@sparted/node/no-extraneous-require': [ERROR],
    '@sparted/node/no-missing-import': [ERROR],
    '@sparted/node/no-missing-require': [ERROR],
    '@sparted/node/no-new-require': [ERROR],
    '@sparted/node/no-path-concat': [ERROR],
    '@sparted/node/no-process-exit': [ERROR],
    '@sparted/node/no-unpublished-bin': [ERROR],
    '@sparted/node/no-unpublished-import': [ERROR],
    '@sparted/node/no-unpublished-require': [ERROR],
    '@sparted/node/no-unsupported-features/es-builtins': [ERROR],
    '@sparted/node/no-unsupported-features/es-syntax': [ERROR],
    '@sparted/node/no-unsupported-features/node-builtins': [ERROR],
    '@sparted/node/process-exit-as-throw': [ERROR],
    '@sparted/node/shebang': [ERROR],

    /* ----- Best practices ----- */
    '@sparted/node/no-deprecated-api': [ERROR, {
      ignoreModuleItems: [], // Default
      ignoreGlobalItems: [], // Default
    }],

    /* ----- Stylistic issues ----- */
    '@sparted/node/callback-return': [ERROR, [
      'callback',
      'cb',
      'next',
      'done',
      'res.send',
      'response.send',
    ]],
    '@sparted/node/exports-style': [ERROR, 'module.exports', {
      allowBatchAssign: false, // Default
    }],
    '@sparted/node/file-extension-in-import': [ERROR, 'always'],
    '@sparted/node/global-require': [ERROR],
    '@sparted/node/no-mixed-requires': [ERROR, {
      grouping: true,
      allowCall: false, // Default
    }],
    '@sparted/node/no-process-env': [ERROR],
    '@sparted/node/no-restricted-import': [ERROR],
    '@sparted/node/no-restricted-require': [ERROR],
    '@sparted/node/no-sync': [ERROR, {
      allowAtRootLevel: false, // Default
    }],
    '@sparted/node/prefer-global/buffer': [ERROR, 'never'],
    '@sparted/node/prefer-global/console': [ERROR, 'never'],
    '@sparted/node/prefer-global/process': [ERROR, 'never'],
    '@sparted/node/prefer-global/text-decoder': [ERROR, 'never'],
    '@sparted/node/prefer-global/text-encoder': [ERROR, 'never'],
    '@sparted/node/prefer-global/url-search-params': [ERROR, 'never'],
    '@sparted/node/prefer-global/url': [ERROR, 'never'],
    '@sparted/node/prefer-promises/dns': [ERROR],
    '@sparted/node/prefer-promises/fs': [ERROR],

    /* ----- Reset base config ----- */
    // TODO: Make a function to merge with original rule config
    // Resolve conflicts with all node/prefer-global/* rules
    'no-shadow': [ERROR, {
      builtinGlobals: true,
      hoist: 'all',
      allow: [
        'Promise',
        'Buffer',
        'console',
        'process',
        'TextDecoder',
        'TextEncoder',
        'URLSearchParams',
        'URL',
      ],
    }],
  },
};
