// @sparted/eslint-plugin/lib/configs/plugins/import/common.js
'use strict';

const process = require('process');


const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';


module.exports = {
  settings: {
    'import/extensions': [
      '.js',
      '.jsx',
      '.json',
      '.json5',
    ],
    'import/ignore': [
      'node_modules',
      'node_modules/react-native/index\\.js$',
    ],
    'import/core-modules': [],
    'import/external-module-folders': [
      'node_modules',
    ],
    'import/parsers': {},
    'import/resolver': {
      'babel-module': {},
    },
    // Might change when using eslint_d
    // @see https://github.com/benmosher/eslint-plugin-import#importcache
    'import/cache': {
      lifetime: Infinity,
    },
    'import/internal-regex': '',
  },
  rules: {
    /* ----- Static analysis ----- */
    '@sparted/import/no-unresolved': [ERROR, {
      commonjs: true,
      amd: false, // Default
      caseSensitive: true, // Default
    }],
    '@sparted/import/named': [ERROR],
    '@sparted/import/default': [ERROR],
    '@sparted/import/namespace': [ERROR, {
      allowComputed: true,
    }],
    // OFF as this rule is project specific
    '@sparted/import/no-restricted-paths': [OFF],
    '@sparted/import/no-absolute-path': [ERROR, {
      esmodule: true, // Default
      commonjs: true, // Default
      amd: false, // Default
    }],
    '@sparted/import/no-dynamic-require': [ERROR],
    // TODO: Might be set to ERROR
    '@sparted/import/no-internal-modules': [OFF, {
      allow: [], // Default
    }],
    '@sparted/import/no-webpack-loader-syntax': [ERROR],
    '@sparted/import/no-self-import': [ERROR],
    '@sparted/import/no-cycle': [ERROR, {
      maxDepth: Infinity, // Default
      commonjs: false, // Default
      amd: false, // Default
      ignoreExternal: false, // Default
    }],
    '@sparted/import/no-useless-path-segments': [ERROR, {
      noUselessIndex: true,
    }],
    // TODO: Check if it works correctly with aliases
    '@sparted/import/no-relative-parent-imports': [ERROR],

    /* ----- Helpful warnings ----- */
    '@sparted/import/export': [ERROR],
    '@sparted/import/no-named-as-default': [ERROR],
    '@sparted/import/no-named-as-default-member': [ERROR],
    '@sparted/import/no-deprecated': [ERROR],
    // TODO: Needs to be modified for test, tools, ...
    '@sparted/import/no-extraneous-dependencies': [ERROR, {
      devDependencies: false,
      optionalDependencies: false,
      peerDependencies: false, // Default
      bundledDependencies: false,
    }],
    '@sparted/import/no-mutable-exports': [ERROR],
    '@sparted/import/no-unused-modules': [ERROR, {
      missingExports: true,
      unusedExports: true,
      src: [process.cwd()], // Default
      // ignoreExports: [], // Default
    }],

    /* ----- Module systems ----- */
    '@sparted/import/unambiguous': [ERROR],
    // Dependent of the project type
    '@sparted/import/no-commonjs': [ERROR, {
      allowRequire: false, // Default
      allowConditionalRequire: true, // Default
      allowPrimitiveModules: false, // Default
    }],
    '@sparted/import/no-amd': [ERROR],
    '@sparted/import/no-nodejs-modules': [ERROR, {
      allow: [], // Default
    }],

    /* ----- Style guide ----- */
    '@sparted/import/first': [ERROR],
    '@sparted/import/exports-last': [ERROR],
    '@sparted/import/no-duplicates': [ERROR, {
      considerQueryString: false, // Default
    }],
    // OFF as it is useful
    '@sparted/import/no-namespace': [OFF],
    '@sparted/import/extensions': [ERROR, 'always', {
      ignorePackages: true,
      pattern: {
        js: 'ignorePackages',
      },
    }],
    '@sparted/import/order': [ERROR, {
      groups: [
        'builtin',
        'external',
        'internal',
        'parent',
        'index',
        'sibling',
        'unknown',
        // Typescript only
        'object',
      ],
      pathGroups: [],
      pathGroupsExcludedImportTypes: [], // Default
      'newlines-between': 'always',
      alphabetize: {
        order: 'ignore', // Default
        caseInsensitive: false, // Default
      },
    }],
    '@sparted/import/newline-after-import': [ERROR, {
      count: 2,
    }],
    '@sparted/import/prefer-default-export': [ERROR],
    // Should be OFF for flow type project
    '@sparted/import/max-dependencies': [ERROR, {
      max: 10, // Default
    }],
    '@sparted/import/no-unassigned-import': [ERROR],
    '@sparted/import/no-named-default': [ERROR],
    // OFF as it conlicts with the rule import/prefer-default-export
    '@sparted/import/no-default-export': [OFF],
    // OFF as named exports are good
    '@sparted/import/no-named-export': [OFF],
    '@sparted/import/no-anonymous-default-export': [ERROR, {
      allowArray: true,
      allowArrowFunction: false, // Default
      allowAnonymousClass: false, // Default
      allowAnonymousFunction: false, // Default
      allowCallExpression: false,
      allowLiteral: false, // Default
      allowObject: true,
    }],
    // OFF as it causes too much constraints on code organization
    '@sparted/import/group-exports': [OFF],
    '@sparted/import/dynamic-import-chunkname': [ERROR, {
      importFunctions: ['import'], // Default
      webpackChunknameFormat: '[a-zA-Z0-9-/_]+', // Default
    }],

    /* ----- Reset base config ----- */
    // Enhanced by the rule import/no-duplicates
    'no-duplicate-imports': [OFF],
  },
};
