// @sparted/eslint-plugin/lib/configs/plugins/react/common.js
'use strict';

const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';

const DISALLOW_IN_FUNC = 'disallow-in-func';
const STATIC_PUBLIC_FIELD = 'static public field';
const PARENS_NEW_LINE = 'parens-new-line';

module.exports = {
  parserOptions: {
    ecmaVersion: 12,
    ecmaFeatures: {
      jsx: true,
    },
  },
  settings: {
    react: {
      createClass: 'createReactClass',
      pragma: 'React',
      fragment: 'Fragment',
      version: 'detect',
      flowVersion: '0.136.0',
    },
    propWrapperFunctions: [],
    linkComponents: [
      { name: 'Link', linkAttribute: 'to' },
    ],
  },
  rules: {
    /* ----- React ----- */
    '@sparted/react/boolean-prop-naming': [ERROR, {
      propTypeNames: ['bool'], // Default
      rule: '^(is|has)[A-Z]([A-Za-z0-9]?)+', // Default
      message: `Prop name ({{ propName }}) doesn't match the rule ({{ pattern }})`, // Default
      validateNested: false, // Default
    }],
    '@sparted/react/button-has-type': [ERROR, {
      button: true, // Default
      submit: true, // Default
      reset: true, // Default
    }],
    '@sparted/react/default-props-match-prop-types': [ERROR, {
      allowRequiredDefaults: false, // Default
    }],
    '@sparted/react/destructuring-assignment': [ERROR, 'always', { // Default
      ignoreClassFields: false, // Default
    }],
    '@sparted/react/display-name': [ERROR, {
      ignoreTranspilerName: false, // Default
    }],
    '@sparted/react/forbid-component-props': [ERROR, {
      forbid: ['className', 'style'], // Default
    }],
    '@sparted/react/forbid-dom-props': [ERROR, {
      forbid: [], // Default
    }],
    '@sparted/react/forbid-elements': [ERROR, {
      forbid: [], // Default
    }],
    '@sparted/react/forbid-foreign-prop-types': [ERROR],
    '@sparted/react/forbid-prop-types': [ERROR, {
      forbid: ['any', 'array', 'object'], // Default
      checkContextTypes: false, // Default
      checkChildContextTypes: false, // Default
    }],
    '@sparted/react/function-component-definition': [ERROR, {
      namedComponents: 'function-declaration', // Default
      unnamedComponents: 'arrow-function',
    }],
    // Shouldn't warn much because of the rule react/destructuring-assignment
    '@sparted/react/no-access-state-in-setstate': [ERROR],
    '@sparted/react/no-adjacent-inline-elements': [ERROR],
    '@sparted/react/no-array-index-key': [ERROR],
    '@sparted/react/no-children-prop': [ERROR],
    '@sparted/react/no-danger': [ERROR],
    '@sparted/react/no-danger-with-children': [ERROR],
    '@sparted/react/no-deprecated': [ERROR],
    '@sparted/react/no-did-mount-set-state': [ERROR, DISALLOW_IN_FUNC],
    '@sparted/react/no-did-update-set-state': [ERROR, DISALLOW_IN_FUNC],
    '@sparted/react/no-direct-mutation-state': [ERROR],
    '@sparted/react/no-find-dom-node': [ERROR],
    '@sparted/react/no-is-mounted': [ERROR],
    '@sparted/react/no-multi-comp': [ERROR, {
      ignoreStateless: false, // Default
    }],
    '@sparted/react/no-redundant-should-component-update': [ERROR],
    '@sparted/react/no-render-return-value': [ERROR],
    // OFF as we use local component state
    '@sparted/react/no-set-state': [OFF],
    '@sparted/react/no-string-refs': [ERROR, {
      noTemplateLiterals: true,
    }],
    '@sparted/react/no-this-in-sfc': [ERROR],
    '@sparted/react/no-typos': [ERROR],
    '@sparted/react/no-unescaped-entities': [ERROR, {
      forbid: ['>', '"', `'`, '}'], // Default
    }],
    '@sparted/react/no-unknown-property': [ERROR, {
      ignore: [], // Default
    }],
    '@sparted/react/no-unsafe': [ERROR, {
      checkAliases: true,
    }],
    '@sparted/react/no-unused-prop-types': [ERROR, {
      customValidators: [], // Default
      skipShapeProps: true, // Default
    }],
    '@sparted/react/no-unused-state': [ERROR],
    '@sparted/react/no-will-update-set-state': [ERROR, DISALLOW_IN_FUNC],
    '@sparted/react/prefer-es6-class': [ERROR, 'always'],
    '@sparted/react/prefer-read-only-props': [ERROR],
    // OFF as we use class components
    '@sparted/react/prefer-stateless-function': [OFF, {
      ignorePureComponents: false, // Default
    }],
    '@sparted/react/prop-types': [ERROR, {
      ignore: [], // Default
      customValidators: [], // Default
      skipUndeclared: false, // Default
    }],
    '@sparted/react/react-in-jsx-scope': [ERROR],
    '@sparted/react/require-default-props': [ERROR, {
      forbidDefaultForRequired: true,
      ignoreFunctionalComponents: false, // Default
    }],
    // OFF as it will add noise before thinking of optimizing
    '@sparted/react/require-optimization': [OFF, {
      allowDecorators: [], // Default
    }],
    '@sparted/react/require-render-return': [ERROR],
    '@sparted/react/self-closing-comp': [ERROR, {
      component: true, // Default
      html: true, // Default
    }],
    '@sparted/react/sort-comp': [ERROR, {
      order: [
        'type-annotations',
        'static-variables',
        'static-methods',
        'instance-variables',
        'lifecycle',
        'rendering',
        'event-handlers',
        'everything-else',
      ],
      groups: {
        rendering: [
          'render',
          '/^render.+$/',
        ],
        'event-handlers': [
          '/^handle.+$/',
        ],
      },
    }],
    '@sparted/react/sort-prop-types': [ERROR, {
      callbacksLast: true,
      ignoreCase: true,
      requiredFirst: true,
      sortShapeProp: true,
      noSortAlphabetically: true,
    }],
    '@sparted/react/state-in-constructor': [ERROR, 'always'],
    // Might be changed
    '@sparted/react/static-property-placement': [ERROR, STATIC_PUBLIC_FIELD, {
      childContextTypes: STATIC_PUBLIC_FIELD,
      contextTypes: STATIC_PUBLIC_FIELD,
      contextType: STATIC_PUBLIC_FIELD,
      defaultProps: STATIC_PUBLIC_FIELD,
      displayName: STATIC_PUBLIC_FIELD,
      propTypes: STATIC_PUBLIC_FIELD,
    }],
    '@sparted/react/style-prop-object': [ERROR, {
      allow: [], // Default
    }],
    '@sparted/react/void-dom-elements-no-children': [ERROR],

    /* ----- JSX ----- */
    '@sparted/react/jsx-boolean-value': [ERROR, 'always'],
    '@sparted/react/jsx-child-element-spacing': [ERROR],
    '@sparted/react/jsx-closing-bracket-location': [ERROR, {
      nonEmpty: 'line-aligned',
      selfClosing: 'line-aligned',
    }],
    '@sparted/react/jsx-closing-tag-location': [ERROR],
    '@sparted/react/jsx-curly-brace-presence': [ERROR, {
      props: 'never',
      children: 'never',
    }],
    // Might be changed to `consistent` for multiline
    '@sparted/react/jsx-curly-newline': [ERROR, {
      multiline: 'require',
      singleline: 'forbid',
    }],
    '@sparted/react/jsx-curly-spacing': [ERROR, {
      when: 'never',
      allowMultiline: true, // Default
      attributes: { when: 'never', allowMultiline: true },
      children: { when: 'never', allowMultiline: true },
      spacing: { objectLiterals: 'never' },
    }],
    '@sparted/react/jsx-equals-spacing': [ERROR, 'never'],
    '@sparted/react/jsx-filename-extension': [ERROR, {
      allow: 'as-needed',
      extensions: ['.jsx'],
    }],
    '@sparted/react/jsx-first-prop-new-line': [ERROR, 'multiline'],
    '@sparted/react/jsx-fragments': [ERROR, 'element'],
    '@sparted/react/jsx-handler-names': [ERROR, {
      eventHandlerPrefix: 'handle', // Default
      eventHandlerPropPrefix: 'on', // Default
      checkLocalVariables: false, // Default
      checkInlineFunction: true,
    }],
    '@sparted/react/jsx-indent': [ERROR, 2, {
      checkAttributes: true,
      indentLogicalExpressions: true,
    }],
    '@sparted/react/jsx-indent-props': [ERROR, 2],
    '@sparted/react/jsx-key': [ERROR, {
      checkFragmentShorthand: true,
      // Will be released in the release following v7.21.5
      // checkKeyMustBeforeSpread: true,
    }],
    '@sparted/react/jsx-max-depth': [ERROR, {
      max: 4,
    }],
    '@sparted/react/jsx-max-props-per-line': [ERROR, {
      maximum: 1, // Default
      when: 'multiline',
    }],
    // OFF as we don't enforce new line between adjacent JSX elements
    '@sparted/react/jsx-newline': [OFF],
    '@sparted/react/jsx-no-bind': [ERROR, {
      ignoreDOMComponents: false, // Default
      ignoreRefs: false, // Default
      allowArrowFunctions: false, // Default
      allowFunctions: false, // Default
      allowBind: false, // Default
    }],
    '@sparted/react/jsx-no-comment-textnodes': [ERROR],
    '@sparted/react/jsx-no-duplicate-props': [ERROR, {
      ignoreCase: true,
    }],
    // TODO: Wait for the version >= 7.21.5
    // '@sparted/react/jsx-no-constructed-context-values': [ERROR],
    '@sparted/react/jsx-no-literals': [ERROR, {
      noStrings: true,
      allowedStrings: [], // Default
      ignoreProps: true,
      noAttributeStrings: false, // Default
    }],
    '@sparted/react/jsx-no-script-url': [ERROR],
    '@sparted/react/jsx-no-target-blank': [ERROR, {
      allowReferrer: false, // Default
      enforceDynamicLinks: 'always',
    }],
    '@sparted/react/jsx-no-undef': [ERROR, {
      allowGlobals: false, // Default
    }],
    '@sparted/react/jsx-no-useless-fragment': [ERROR],
    '@sparted/react/jsx-one-expression-per-line': [ERROR, {
      allow: 'none', // Default
    }],
    '@sparted/react/jsx-pascal-case': [ERROR, {
      allowAllCaps: false, // Default
      ignore: [], // Default
    }],
    '@sparted/react/jsx-props-no-multi-spaces': [ERROR],
    // OFF as it doesn't allow HOC
    '@sparted/react/jsx-props-no-spreading': [OFF, {
      html: 'enforce', // Default
      custom: 'enforce', // Default
      explicitSpread: 'enforce', // Default
      exceptions: [], // Default
    }],
    // OFF as it is a nightmare to maintain for little benefit
    '@sparted/react/jsx-sort-default-props': [OFF, {
      ignoreCase: true,
    }],
    '@sparted/react/jsx-sort-props': [ERROR, {
      callbacksLast: true,
      shorthandFirst: true,
      shorthandLast: false, // Default
      ignoreCase: true,
      noSortAlphabetically: true,
      reservedFirst: true,
    }],
    '@sparted/react/jsx-tag-spacing': [ERROR, {
      closingSlash: 'never', // Default
      beforeSelfClosing: 'always', // Default
      afterOpening: 'never', // Default
      beforeClosing: 'never',
    }],
    '@sparted/react/jsx-uses-react': [ERROR],
    '@sparted/react/jsx-uses-vars': [ERROR],
    '@sparted/react/jsx-wrap-multilines': [ERROR, {
      declaration: PARENS_NEW_LINE,
      assignment: PARENS_NEW_LINE,
      return: PARENS_NEW_LINE,
      arrow: PARENS_NEW_LINE,
      condition: PARENS_NEW_LINE,
      logical: PARENS_NEW_LINE,
      prop: 'ignore', // Default
    }],
  },
};
