// eslint-plugin-listeners/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDescription = exports.parseMemberExpression = exports.isNodeArrowFunctionExpression = exports.isNodeFunctionExpression = exports.isNodeThisExpression = exports.isNodeMemberExpression = exports.isNodeIdentifier = exports.RuleType = void 0;
var RuleType;
(function (RuleType) {
    RuleType["MissingRemoveEventListener"] = "no-missing-remove-event-listener";
    RuleType["MatchingRemoveEventListener"] = "matching-remove-event-listener";
    RuleType["InlineFunctionEventListener"] = "no-inline-function-event-listener";
})(RuleType = exports.RuleType || (exports.RuleType = {}));
exports.isNodeIdentifier = (node) => (node === null || node === void 0 ? void 0 : node.type) === 'Identifier';
exports.isNodeMemberExpression = (node) => (node === null || node === void 0 ? void 0 : node.type) === 'MemberExpression';
exports.isNodeThisExpression = (node) => (node === null || node === void 0 ? void 0 : node.type) === 'ThisExpression';
exports.isNodeFunctionExpression = (node) => (node === null || node === void 0 ? void 0 : node.type) === 'FunctionExpression';
exports.isNodeArrowFunctionExpression = (node) => (node === null || node === void 0 ? void 0 : node.type) === 'ArrowFunctionExpression';
exports.parseMemberExpression = (node) => {
    let value;
    if (exports.isNodeIdentifier(node.object)) {
        value = node.object.name;
    }
    if (exports.isNodeMemberExpression(node.object)) {
        value = exports.parseMemberExpression(node.object);
    }
    if (exports.isNodeThisExpression(node.object)) {
        value = `this.${node.property.name}`;
    }
    return value;
};
exports.getDescription = (ruleName) => {
    switch (ruleName) {
        case RuleType.MissingRemoveEventListener:
            return 'No missing removeEventListener if addEventListener exists';
        case RuleType.MatchingRemoveEventListener:
            return 'Handler for removeEventListener should match addEventListener';
        case RuleType.InlineFunctionEventListener:
            return 'Handler for event listener should not be an inline function';
    }
};
